crates/sui-metrics-proxy/src/admin.rs

```diff
@@ -0,0 +1,163 @@
+    hostname: &str,
+) -> Result<(ServerConfig, ValidatorAllowlist), sui_tls::rustls::Error> {
+    let mut rng = rand::thread_rng();
+    let server_keypair = Ed25519KeyPair::generate(&mut rng);
+    let server_certificate = SelfSignedCertificate::new(server_keypair.private(), hostname);
+
+    ValidatorCertVerifier::rustls_server_config(
+        vec![server_certificate.rustls_certificate()],
+        server_certificate.rustls_private_key(),
+    )
+}
+
+async fn get_validators(url: String) -> Result<Vec<ValidatorMetadata>> {
+    let client = reqwest::Client::builder().build().unwrap();
+    let request = serde_json::json!({
+        "jsonrpc": "2.0",
+        "method":"sui_getValidators",
+        "id":1,
+    });
+    let response = client
+        .post(url)
+        .header(reqwest::header::CONTENT_TYPE, "application/json")
+        .body(request.to_string())
+        .send()
+        .await
+        .context("unable to perform rpc")?;
+
+    #[derive(Debug, Deserialize)]
+    struct ResponseBody {
+        result: Vec<ValidatorMetadata>,
+    }
+    let body = response
+        .json::<ResponseBody>()
+        .await
+        .context("unable to deserialize validator peer list")?;
+
+    Ok(body.result)
+}
+
+pub fn manage_validators(url: String, period: Duration, allowlist: ValidatorAllowlist) {
+    info!("Started polling for peers using rpc: {}", url);
+    tokio::spawn(async move {
+        let mut interval = tokio::time::interval(period);
+        interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+
+        loop {
+            interval.tick().await;
+
+            match get_validators(url.clone()).await {
+                Ok(peers) => {
+                    // obtain rwlock; nb allowed is an raii object
+                    let mut allowed = allowlist
+                        .write()
+                        // TODO is expect ok here? write can panic in some cases per docs
+                        .expect("unable to obtain rwlock for peer list update");
+                    // empty our list, may result in an empty allow list if len(peers) == 0
+                    allowed.clear();
+                    for v in peers {
+                        info!("name: {:?} sui_address: {:?}", v.name, v.sui_address);
+                        match Ed25519PublicKey::from_bytes(&v.network_pubkey_bytes) {
+                            Ok(client_public_key) => {
+                                // Insert the client's public key into the allowlist and verify the request is successful
+                                allowed.insert(client_public_key);
+                            }
+                            Err(error) => {
+                                error!(
+                                    "unable to decode public key for name: {:?} sui_address: {:?} error: {error}",
+                                    v.name, v.sui_address);
+                            }
+                        }
+                    }
```

Comment:
i like this!  Prior to this loop, i purge the set and then re-add values.  This will let me swap them out using replace.  https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.replace

thanks! not sure why i wrote it this way.  I'm used to the imperative style more.  i need to think about this iterator pattern as a first approach more often. thanks.